tools/clang/unittests/HLSL/OptimizerTest.cpp

```diff
@@ -259,3 +303,1152 @@ void OptimizerTest::OptimizerWhenSliceNThenOK(int optLevel, LPCSTR pText, LPCWST
+      originalContainer, hlsl::DFCC_PipelineStateValidation);
+
+  auto optimizedContainer = RunHlslEmitAndReturnContainer(originalContainer);
+
+  auto optimizedPsvPart = RetrievePartFromContainer(optimizedContainer,
+                                           hlsl::DFCC_PipelineStateValidation);
+
+  VERIFY_ARE_EQUAL(originalPsvPart->GetBufferSize(), optimizedPsvPart->GetBufferSize());
+
+  VERIFY_ARE_EQUAL(memcmp(originalPsvPart->GetBufferPointer(),
+                          optimizedPsvPart->GetBufferPointer(),
+                          originalPsvPart->GetBufferSize()),
+                   0);
+
+  DxilPipelineStateValidation originalPsv;
+  originalPsv.InitFromPSV0(
+      reinterpret_cast<const uint8_t *>(optimizedPsvPart->GetBufferPointer()),
+      static_cast<uint32_t>(optimizedPsvPart->GetBufferSize()));
+
+  const PSVRuntimeInfo1 *originalInfo1 = originalPsv.GetPSVRuntimeInfo1();
+  if (usesViewId) {
+    VERIFY_IS_TRUE(originalInfo1->UsesViewID);
+  }
+
+  DxilPipelineStateValidation optimizedPsv;
+  optimizedPsv.InitFromPSV0(
+      reinterpret_cast<const uint8_t *>(originalPsvPart->GetBufferPointer()),
+      static_cast<uint32_t>(originalPsvPart->GetBufferSize()));
+
+  const PSVRuntimeInfo1 *optimizedInfo1 = optimizedPsv.GetPSVRuntimeInfo1();
+
+  VERIFY_ARE_EQUAL(originalInfo1->ShaderStage, optimizedInfo1->ShaderStage);
+  VERIFY_ARE_EQUAL(originalInfo1->UsesViewID , optimizedInfo1->UsesViewID);
+  VERIFY_ARE_EQUAL(originalInfo1->SigInputElements,
+                   optimizedInfo1->SigInputElements);
+  VERIFY_ARE_EQUAL(originalInfo1->SigOutputElements,
+                   optimizedInfo1->SigOutputElements);
+  VERIFY_ARE_EQUAL(originalInfo1->SigPatchConstOrPrimElements,
+                   optimizedInfo1->SigPatchConstOrPrimElements);
+  VERIFY_ARE_EQUAL(originalPsv.GetViewIDPCOutputMask().IsValid(),
+                   optimizedPsv.GetViewIDPCOutputMask().IsValid());
+  if (originalPsv.GetViewIDPCOutputMask().IsValid()) {
+    VERIFY_ARE_EQUAL(*originalPsv.GetViewIDPCOutputMask().Mask,
+                     *optimizedPsv.GetViewIDPCOutputMask().Mask);
+    VERIFY_ARE_EQUAL(originalPsv.GetViewIDPCOutputMask().NumVectors,
+                     optimizedPsv.GetViewIDPCOutputMask().NumVectors);
+  }
+  for (int stream = 0; stream < streamCount; ++stream) 
+  {
+    VERIFY_ARE_EQUAL(originalInfo1->SigOutputVectors[stream],
+                     optimizedInfo1->SigOutputVectors[stream]);
+
+    VERIFY_ARE_EQUAL(originalPsv.GetViewIDOutputMask(stream).IsValid(),
+                     optimizedPsv.GetViewIDOutputMask(stream).IsValid());
+    if (originalPsv.GetViewIDOutputMask(stream).IsValid())
+    {
+      VERIFY_ARE_EQUAL(*originalPsv.GetViewIDOutputMask(stream).Mask,
+                       *optimizedPsv.GetViewIDOutputMask(stream).Mask);
+      VERIFY_ARE_EQUAL(originalPsv.GetViewIDOutputMask(stream).NumVectors,
+                       optimizedPsv.GetViewIDOutputMask(stream).NumVectors);
+
+    }
```

Comment:
This misses comparing the input->output dependency bits, which are set regardless of whether `SV_ViewID` is used in the shader.

Reply:
LMK if what I've done is insufficient.

Reply:
Number of input and output vectors is not going to be different as long as there is data here, it's the contents of the masks that matter.  If you assume that the mere presence of the ViewID data is enough, this check is enough, but I'm not sure that's enough given the potential to make a mistake in the code that transfers the data from PSV0 to the DxilModule serialized format.

Reply:
A simple comparison of the data might be fine.  The data at `PSVDependencyTable::Table` is an array of `uint32_t` with an array size of `PSVComputeMaskDwordsFromVectors(OutputVectors) * InputVectors * 4`.  You could just to a `memcmp` on that.

Reply:
This also misses any potential table from `GetInputToPCOutputTable` or `GetPCInputToOutputTable`.  The content `memcmp` could be performed when `InputVectors * OutputVectors > 0`.

Likely it'll be best to write a little helper function (or lambda), like this:
```
void VerifyIOTablesMatch(const PSVDependencyTable &original,
                         const PSVDependencyTable &optimized) {
  VERIFY_ARE_EQUAL(original.InputVectors, optimized.InputVectors);
  VERIFY_ARE_EQUAL(original.OutputVectors, optimized.OutputVectors);
  if (original.InputVectors == optimized.InputVectors &&
      original.OutputVectors == optimized.OutputVectors &&
      original.InputVectors * original.OutputVectors > 0) {
    VERIFY_ARE_EQUAL(
        0, memcmp(original.Table, optimized.Table,
                  PSVComputeMaskDwordsFromVectors(original.OutputVectors) *
                      original.InputVectors * 4));
  }
}
```

Then use that on all three possible tables.

Reply:
That won't tell you where it went wrong, but will catch it if it is wrong.

Reply:
Now testing all 3 tables